### PR TITLE
8295698: AArch64: test/jdk/sun/security/ec/ed/EdDSATest.java failed with -XX:+UseSHA3Intrinsics

### PR DESCRIPTION
**Pull request by dgbo**
In JDK-8252204, when implemented SHA3 intrinsics, we use `digest_length` to differentiate SHA3-224, SHA3-256, SHA3-384, SHA3-512 and calculate `block_size` with `block_size = 200 - 2 * digest_length`.
However, there are two extra SHA3 instances, SHAKE256 and SHAKE128, allowing an arbitrary `digest_length`:
```
	digest_length	block_size
SHA3-224	28	144
SHA3-256	32	136
SHA3-384	48	104
SHA3-512	64	72
SHAKE128	variable	168
SHAKE256	variable	136
```

This causes SIGSEGV crash or hash code mismatch with `test/jdk/sun/security/ec/ed/EdDSATest.java`. The test calls `SHAKE256` in `Ed448`.

The main idea of the patch is to pass the `block_size` to differentiate SHA3 instances.
Tests `test/jdk/sun/security/ec/ed/EdDSATest.java` and `./test/jdk/sun/security/provider/MessageDigest/SHA3.java` both passed.
And tier1~3 passed on SHA3 supported hardware.

The SHA3 intrinsics still deliver 20%~40% performance improvement on our pre-silicon simulated platform.
The latency and throughput of crypto SHA3 ops are designed to be 1 cpu cycle and 2 execution pipes respectively.

Compared with the main stream code, the performance change with this patch are negligible on real hardware and simulation platform.
Based on the JMH results of SHA3 intirinsics, performance can be improved by ~50% on some hardware, while some hardware have ~30% regression.
These performance details are available in the comments of the issue page.
I guess the performance benefit of SHA3 intrinsics is dependent on the micro architecture, it should be switched on/off based on the running platform.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8295698](https://bugs.openjdk.org/browse/JDK-8295698): AArch64: test/jdk/sun/security/ec/ed/EdDSATest.java failed with -XX:+UseSHA3Intrinsics

### Reviewers
 * [Hao Sun](https://openjdk.org/census#haosun) (@shqking - Author)
 * [Andrew Haley](https://openjdk.org/census#aph) (@theRealAph - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10939/head:pull/10939` \
`$ git checkout pull/10939`

Update a local copy of the PR: \
`$ git checkout pull/10939` \
`$ git pull https://git.openjdk.org/jdk pull/10939/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10939`

View PR using the GUI difftool: \
`$ git pr show -t 10939`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10939.diff">https://git.openjdk.org/jdk/pull/10939.diff</a>

</details>

### CODE REVIEW

**Review by shqking — APPROVED** [review]
Overall, it looks good to me. (I'm not a Reviewer)

I agree with your fix to pass `block_size` rather than `digest_length`, but I didn't fully understand your update insides `stubGenerator_aarch64.cpp`, i.e. those operations before `rounds24_loop`, since I'm not a crypto expert.
We may need some crypo expert to review that part.

Note that I rerun tier1~3 with the latest commit in this PR on sha512 feature supported hardware, and the test passed.

Thanks.

**Comment by dgbo** [comment]
Thanks for the review.

Before the `NR (==24)` iterations in `keccak`, `a0~a24` need to be ready:
```
// SHA3.java
    private void keccak() {
        // convert the 200-byte state into 25 lanes
        bytes2Lanes(state, lanes);
        ...
        a0 = lanes[0]; a1 = lanes[1]; ...; a24 = lanes[24];

        // process the lanes through step mappings
        for (int ir = 0; ir < NR; ir++) {
            long c0 = a0^a5^a10^a15^a20;
            long c1 = a1^a6^a11^a16^a21;
            ...
```

While `a_i` is computed as: `state[i] ^= b[ofs++]` in implCompress0(byte[] b, int ofs), `byte2Lanes(state, lanes)`, `a_i = lanes[i]`.
```
// SHA3.java
    @IntrinsicCandidate
    private void implCompress0(byte[] b, int ofs) {
       for (int i = 0; i < buffer.length; i++) {
           state[i] ^= b[ofs++];
       }
       keccak();
    }
```

Those operations before `rounds24_loop` can be taken as vectorial version of the Java code shown above:
1. Load all elements of `state` array into vector registers (v0~v24).
2. Load the input data into `v25~v31`. The number of data to load is dependent on `block_size (== buffer.length)`.
3. Perform `state[i] ^ b[ofs]`.
We don't need instructions for `byte2Lanes`, cause it is already done with the `ld1_8B` instructions.

Hope this can address your confusion.

**Comment by shqking** [comment]
Thanks for your kind explanation. Understood.
This patch is good to me. (I'm not a Reviewer)

**Comment by dgbo** [comment]
Hi, @nick-arm @theRealAph, could you help to review this? Thanks.

**Comment by theRealAph** [comment]
This looks right, but I don't think I can test it, which I usually would do with a patch this complicated. When we have a processor without FEAT_SHA3) we should define BCAX, EOR3, RAX1, and XAR as macros. Could you do that, please?

**Comment by dgbo** [comment]
> This looks right, but I don't think I can test it, which I usually would do with a patch this complicated. When we have a processor without FEAT_SHA3) we should define BCAX, EOR3, RAX1, and XAR as macros. Could you do that, please?

Thanks for the comments.

Do you mean that we need a patch (not to be merged) to support testing on processor without FEAT_SHA3?
In which, the SHA3 instructions are substituted by multiple instructions, something like: `eor3 v1, v2, v3, v4 => eor v1, v2, v3; eor v1, v1, v4`?

BTW, FEAT_SHA3 is supported on M1. If you happen to have one, the test can be done on it. :)
To test this on M1/MacOS, modification below is needed to enable SHA3Intriniscs by default.
Since other features, i.e. UseSHA, can not be automatically detected neither, I think it is irrelevant with this patch.
```
--- a/src/hotspot/cpu/aarch64/vm_version_aarch64.cpp
+++ b/src/hotspot/cpu/aarch64/vm_version_aarch64.cpp
@@ -334,15 +334,15 @@ void VM_Version::initialize() {
     FLAG_SET_DEFAULT(UseSHA256Intrinsics, false);
   }

- if (UseSHA && VM_Version::supports_sha3()) {
+ // if (UseSHA && VM_Version::supports_sha3()) {
     // Do not auto-enable UseSHA3Intrinsics until it has been fully tested on hardware
- // if (FLAG_IS_DEFAULT(UseSHA3Intrinsics)) {
- // FLAG_SET_DEFAULT(UseSHA3Intrinsics, true);
- // }
- } else if (UseSHA3Intrinsics) {
- warning("Intrinsics for SHA3-224, SHA3-256, SHA3-384 and SHA3-512 crypto hash functions not available on this CPU.");
- FLAG_SET_DEFAULT(UseSHA3Intrinsics, false);
- }
+ if (FLAG_IS_DEFAULT(UseSHA3Intrinsics)) {
+ FLAG_SET_DEFAULT(UseSHA3Intrinsics, true);
+ }
+ //} else if (UseSHA3Intrinsics) {
+ // warning("Intrinsics for SHA3-224, SHA3-256, SHA3-384 and SHA3-512 crypto hash functions not available on this CPU.");
+ // FLAG_SET_DEFAULT(UseSHA3Intrinsics, false);
+ //}
```

**Comment by theRealAph** [comment]
> > This looks right, but I don't think I can test it, which I usually would do with a patch this complicated. When we have a processor without FEAT_SHA3) we should define BCAX, EOR3, RAX1, and XAR as macros. Could you do that, please?
> 
> Thanks for the comments.
> 
> Do you mean that we need a patch (not to be merged) to support testing on processor without FEAT_SHA3? In which, the SHA3 instructions are substituted by multiple instructions, something like: `eor3 v1, v2, v3, v4 => eor v1, v2, v3; eor v1, v1, v4`?

Yes, exactly. So the intrinsic will work everywhere. Why not merge it? It seems obvious to me.

> BTW, FEAT_SHA3 is supported on M1. If you happen to have one, the test can be done on it. :) To test this on M1/MacOS, modification below is needed to enable SHA3Intriniscs by default. Since other features, i.e. UseSHA, can not be automatically detected neither, I think it is irrelevant with this patch.

Do you know why SHA3 isn't enabled on M1?

**Comment by dgbo** [comment]
> > > This looks right, but I don't think I can test it, which I usually would do with a patch this complicated. When we have a processor without FEAT_SHA3) we should define BCAX, EOR3, RAX1, and XAR as macros. Could you do that, please?
> > 
> > 
> > Thanks for the comments.
> > Do you mean that we need a patch (not to be merged) to support testing on processor without FEAT_SHA3? In which, the SHA3 instructions are substituted by multiple instructions, something like: `eor3 v1, v2, v3, v4 => eor v1, v2, v3; eor v1, v1, v4`?
> 
> Yes, exactly. So the intrinsic will work everywhere. Why not merge it? It seems obvious to me.
>
I see, great idea. But I'm afraid this cannot easily be done due to the high register pressure of the `keccak()` algorithm loop.

Three registers are passed to `xar Vd, Vn, Vm, #imm`, which operation equals to `tmp = Vn ROR Vm; Vd = ROR(tmp[127:64], imm):ROR(tmp[63:0], imm)`.
Registers `Vm/Vn` will be read by operations below, we need another `tmp` register, or we will get clobbered `Vm/Vn` when translating `xar` to NEON instructions.
And we cannot find a free `tmp` register, almost every register has useful data for future use.

The other three SHA3 instructions have similar register issue.
 
> > BTW, FEAT_SHA3 is supported on M1. If you happen to have one, the test can be done on it. :) To test this on M1/MacOS, modification below is needed to enable SHA3Intriniscs by default. Since other features, i.e. UseSHA, can not be automatically detected neither, I think it is irrelevant with this patch.
> 
> Do you know why SHA3 isn't enabled on M1?

FEAT_SHA3 is not detected on MacOS, this code snippet below would be enough for FEAT_SHA3.
I think there are some other hardware features, e.g. FEAT_SHA1, FEAT_SHA256, FEAT_SHA512, are not detected on MacOS too.
We may need another PR to fix it.
```
diff --git a/src/hotspot/os_cpu/bsd_aarch64/vm_version_bsd_aarch64.cpp b/src/hotspot/os_cpu/bsd_aarch64/vm_version_bsd_aarch64.cpp
index 45cd77b3ba5..0ac6b4a56d1 100644
--- a/src/hotspot/os_cpu/bsd_aarch64/vm_version_bsd_aarch64.cpp
+++ b/src/hotspot/os_cpu/bsd_aarch64/vm_version_bsd_aarch64.cpp
@@ -68,6 +68,8 @@ void VM_Version::get_os_cpu_info() {
   if (cpu_has("hw.optional.armv8_crc32"))     _features |= CPU_CRC32;
   if (cpu_has("hw.optional.armv8_1_atomics")) _features |= CPU_LSE;
 
+  if (cpu_has("hw.optional.armv8_2_sha3")) _features |= CPU_SHA3;
+
   int cache_line_size;
   int hw_conf_cache_line[] = { CTL_HW, HW_CACHELINE };
   sysctllen = sizeof(cache_line_size);
```

**Comment by theRealAph** [comment]
Hmm, okay. Looks like there's work to do on this. I'll approve this patch, but we really must get MacOS fixed for JDK 20.

**Review by theRealAph — APPROVED** [review]

**Comment by dgbo** [comment]
> Hmm, okay. Looks like there's work to do on this. I'll approve this patch, but we really must get MacOS fixed for JDK 20.

Thanks for the review. For MacOS issue, I have filed an issue, see https://bugs.openjdk.org/browse/JDK-8297092.
Because we only have a limited testing environment for MacOS, I've asked @shqking for help, he is willing to fix it.

We have two reviews for this SHA3 PR, and it has been fully tested on several real processors.
I am going to integrate this, @theRealAph, would you like to sponsor this? Thanks.

File: src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp [hunk 1]
```diff
@@ -3888,46 +3888,54 @@ class StubGenerator: public StubCodeGenerator {
...
-    __ b(rounds24_loop);
+    __ ld1(v25, v26, v27,  __ T8B, __ post(buf, 24));
+    __ eor(v14, __ T8B, v14, v25);
+    __ eor(v15, __ T8B, v15, v26);
+    __ eor(v16, __ T8B, v16, v27);
 
-    __ BIND(sha3_384_or_224);
-    __ tbz(digest_length, 2, rounds24_loop); // bit 2 cleared? SHA-384
+    // block_size == 136, bit4 == 0 and bit5 == 0, SHA3-256 or SHAKE256
+    __ andw(c_rarg5, block_size, 48);
```
Comment (mur47x111) [inline src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:3910]: does `c_rarg5` serve as a temporary register here?

Comment (dgbo) [reply src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:3910]: Yes, it is a save-on-call register. If it was active, it should been saved.
Similar usages can be found at other intrinsics, such as https://github.com/openjdk/jdk/blob/ab821aa24f248e042d367ccd908fc1f68ebe8333/src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp#L3468.

**Pull request closed without merging**